### PR TITLE
Dice roller

### PR DESCRIPTION
**Pull request by ScrumpySipper**
Initial Implementation of Dice Roller. I can remake this request if you make a new branch for it so I don't have to target main.

### CODE REVIEW

File: main/modes/mode_diceroller.c [hunk 1]
```diff
@@ -0,0 +1,635 @@
...
+#include "swadge_util.h"
+#include "text_entry.h"
+
+#include "mode_diceroller.h"
+
+
+//TODO: Add Doxygen Blocks
+//TODO: Add Cosmetic Features of Colored background/foreground objects. Add Smoother animations. Add LED interaction.
+
+const double pi = 3.14159265359;
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:36]: This is in `math.h` as `M_PI`

File: main/modes/mode_diceroller.c [hunk 2]
```diff
@@ -0,0 +1,635 @@
...
+void changeDiceSidesRequest(int change);
+void doRoll(int count, int sides, int index);
+void doStateMachine(int64_t elapsedUs);
+
+const int MAXDICE = 6;
+const int COUNTCOUNT = 7;
+int validSides[] = {4, 6, 8, 10, 12, 20, 100};
+int polygonSides[] = {3, 4, 3, 4, 5, 3, 6};
+
+int64_t rollAnimationPeriod = 1000000; //1 Second Spin
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:58]: Try to stick to `int32_t` unless really necessary.

File: main/modes/mode_diceroller.c [hunk 3]
```diff
@@ -0,0 +1,635 @@
...
+void drawRegularPolygon(int xCenter, int yCenter, int8_t sides, float rotDeg, int16_t radius, paletteColor_t col, int dashWidth);
+void changeActiveSelection();
+void changeDiceCountRequest(int change);
+void changeDiceSidesRequest(int change);
+void doRoll(int count, int sides, int index);
+void doStateMachine(int64_t elapsedUs);
+
+const int MAXDICE = 6;
+const int COUNTCOUNT = 7;
+int validSides[] = {4, 6, 8, 10, 12, 20, 100};
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:55]: Any variables not in `diceRoller_t` should be `const`

File: main/modes/mode_diceroller.c [hunk 4]
```diff
@@ -0,0 +1,635 @@
...
+    diceRoller->activeSelection = 0;
+
+    diceRoller->state = DR_STARTUP;
+    diceRoller->stateAdvanceFlag = 0;
+}
+
+void diceExitMode(void)
+{
+    freeFont(&diceRoller->ibm_vga8);
+    free(diceRoller->rolls);
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:149]: It feels a little odd to have this freed here, but not allocated in `diceEnterMode`. Are you sure it's allocated? Maybe check that it's not `NULL` first?

File: main/modes/mode_diceroller.c [hunk 5]
```diff
@@ -0,0 +1,635 @@
...
+    DR_SHOWROLL = 1,
+    DR_ROLLING = 2
+};
+
+//int polygonOuterSides[] = []
+//Consider adding outer geometry of dice to make them more recognizable
+
+swadgeMode modeDiceRoller =
+{
+    .modeName = "Dice Roller",
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:74]: `"Dice Roller"` is used a few places, so it should be a `const char[]` above.

File: main/modes/mode_diceroller.c [hunk 6]
```diff
@@ -0,0 +1,635 @@
...
+
+void doStateMachine(int64_t elapsedUs)
+{
+    switch(diceRoller->state)
+    {
+        case DR_STARTUP:
+        {
+                diceRoller->disp->clearPx();
+
+                char staticText[] = "Dice Roller";
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:246]: Another usage of `"Dice Roller"`

File: main/modes/mode_diceroller.c [hunk 7]
```diff
@@ -0,0 +1,635 @@
...
+    for(int m=0; m<count; m++)
+    {
+        diceRoller->rolls[m] = (esp_random() % sides) + 1;
+    }
+    diceRoller->rollSize = count;
+    diceRoller->rollSides = sides;
+    diceRoller->rollIndex = ind;
+}
+
+double cosDeg(double degrees)
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:581]: If you want to move away from `double`, and also expensive trig functions, we have trig functions based on LUTs that return a sin value between -1024 and 1024:
```C
int16_t getSin1024(int16_t degree);
int16_t getCos1024(int16_t degree);
int32_t getTan1024(int16_t degree);
```

File: main/modes/mode_diceroller.c [hunk 8]
```diff
@@ -0,0 +1,635 @@
...
+            xCenter + vertices[endInd].x, yCenter + vertices[endInd].y,
+            col, dashWidth
+            );
+        
+    }
+    
+    free(vertices);
+}
+
+void maskDiceTexture();
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.c:627]: Future functions?

File: main/modes/mode_diceroller.h [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#ifndef MODES_MODE_DICEROLLER_H
+#define MODES_MODE_DICEROLLER_H
+
+extern swadgeMode modeDiceRoller;
+
+typedef struct
+{
+    int16_t x;
+    int16_t y;
+} coord_t;
```
Comment (AEFeinstein) [inline main/modes/mode_diceroller.h:10]: `aabb_utils.h` has the similar:
```C
typedef struct
{
    int32_t x;
    int32_t y;
} vector_t;
```

**Review by AEFeinstein — APPROVED** [review]

File: main/modes/mode_diceroller.c (hunk 1 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.c:36]: Fixed inability to use M_PI by defining _USE_MATH_DEFINES

File: main/modes/mode_diceroller.c (hunk 4 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.c:149]: Initialized pointer to NULL and made free conditional on exit.

File: main/modes/mode_diceroller.c (hunk 5 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.c:74]: Changed to const char[]  in all instances in mode_diceroller.c

File: main/modes/mode_diceroller.c (hunk 3 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.c:55]: Changed variables to const

File: main/modes/mode_diceroller.h (hunk 9 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.h:10]: Switched coord_t for vector_t

File: main/modes/mode_diceroller.c (hunk 2 above)
Comment (ScrumpySipper) [reply main/modes/mode_diceroller.c:58]: limited int64_t usage to esp_clock timestamps

**Pull request merged**